### PR TITLE
Broadcast cancellation to only nodes have outstanding child tasks

### PR DESCRIPTION
**Pull request by dnhatn**
Today when canceling a task we broadcast ban/unban requests to all nodes in the cluster. This strategy does not scale well for hierarchical cancellation. With this change, we will track outstanding child requests and broadcast cancellation to only nodes having outstanding child tasks. This change also strengthens the coordination during cancellation. A parent task will no longer be able to send child requests via transport service once it gets cancelled, which streamlines cancellation on the parent task by not requiring to manually check whether the task is cancelled during the execution.

Relates #50990
Supersedes #51157

Co-authored-by: Igor Motov <igor@motovs.org>
Co-authored-by: Yannick Welsch <yannick@welsch.lu>

### CODE REVIEW

**Comment by dnhatn** [comment]
Please hold off the review. I am investigating some test failures that are related to this change.

**Comment by ywelsch** [comment]
I wonder if, as part of this iteration, we should already wait with the unbanning until all child tasks have been completed.

**Comment by dnhatn** [comment]
> I wonder if, as part of this iteration, we should already wait with the unbanning until all child tasks have been completed (successfully or not).

++. I will add it to this PR.

**Comment by dnhatn** [comment]
This is ready for reviews :).

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: perhaps `runnable = {}`, which avoids the null check later down

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: why does this deserve a warning?

When do we expect this to happen?

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: Why use empty set here and not the actual set? Will this not lead to incorrectness?

File: server/src/main/java/org/elasticsearch/transport/TransportService.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java]: I think that we need to actually capture the connection (not the DiscoveryNode object). The reason for this is cross-cluster connections. In the task cancellation action, we assume that the DiscoveryNode is sufficient to resolve the target of the cancellation. This only works for nodes in the same cluster, however.

I don't think that we need to solve this all in this PR, but we need to capture the issue properly, and add an item to the planning.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java]: should we call this in a finally block?

File: server/src/main/java/org/elasticsearch/transport/TransportService.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java]: should we call this in a finally block?

File: server/src/test/java/org/elasticsearch/search/SearchCancellationIT.java
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/search/SearchCancellationIT.java]: assert that `cancelFuture.isDone == false` before disabling blocks?

File: server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CancellableTasksIT.java]: `testBanOnly...`

**Review by ywelsch — COMMENTED** [review]
Thanks Nhat, I've left some questions and comments.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: Ok, I will try to implement this in a follow-up if it's simple.

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: I restored and added a comment https://github.com/elastic/elasticsearch/pull/54312/commits/297b78726aeb4c7cae96db54e98903174745fd0b

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: I don't know either. It's from the draft version. Sorry.

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: `onChildTasksCompleted` is still null if we do not cancel the task.

File: server/src/main/java/org/elasticsearch/tasks/TaskManager.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/tasks/TaskManager.java]: Good catch. Fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/226a5414320e8e4f725a1762e2226af6731a906a

File: server/src/main/java/org/elasticsearch/transport/TransportService.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: ++ see https://github.com/elastic/elasticsearch/pull/54312/commits/4f337decce08edd355621c9982559df34bc42ce6

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: We throw exception at the end. Anyway, I moved up that statement right after we cancel (see https://github.com/elastic/elasticsearch/pull/54312/commits/a5af596d79d4c663cb711f60f0f1011ad22b32d0).

**Comment by dnhatn** [comment]
@ywelsch Thanks for the reviews. I think I've addressed your comments:
- Fix logging issues in https://github.com/elastic/elasticsearch/pull/54312/commits/03ce7a816e1b89712ebd4057d3550646c315af2e
- Add the wait_for_completion parameter and revert test changes in https://github.com/elastic/elasticsearch/pull/54312/commits/a5af596d79d4c663cb711f60f0f1011ad22b32d0

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks Nhat, I've left some more comments to consider

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: > Another issue I have with this is that this does not play nicely with wait_for_completion. Assume someone else has cancelled the task, and you now want to wait for completion. This gives you a hard exception instead of waiting for all child tasks to complete.

I thought about that scenario but decided not to implement it as I worried about its complexity (we need to keep track of a list of cancellation listeners and childTasksCompleted listeners). Anyway, I will work on it.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: Let's add a TODO for that part, and consider it in a follow-up

**Review by javanna — COMMENTED** [review]
I left a couple of comments. I am glad to see these improvements made to the tasks framework! 

I think it would be valuable to expand a bit the description of this PR and give it a bit more visibiliry. I find that besides no longer broadcasting bans to all nodes, this PR adds more control to the node where the parent task runs. Today, the parent task has to make sure to stop doing its job once it gets cancelled. With this PR the parent task will no longer be able to send child requests via transport service once it gets cancelled, which streamlines cancellation on the parent task by not requiring to manually check whether the task is cancelled during the execution, ad-hoc This is a great improvement, that search will certainly benefit from.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: thanks, fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/a01325e069ff433165a0bdbc0da2854da8db729e

File: rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json
Comment (dnhatn) [reply rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json]: ++ I pushed https://github.com/elastic/elasticsearch/pull/54312/commits/8cff3b0a4998ef4f5de015c484c2f4389a807ce3.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java]: yes, fixed in https://github.com/elastic/elasticsearch/pull/54312/commits/8cff3b0a4998ef4f5de015c484c2f4389a807ce3.

File: x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchActionTests.java
Comment (dnhatn) [reply x-pack/plugin/async-search/src/test/java/org/elasticsearch/xpack/search/AsyncSearchActionTests.java]: Yes, but this is an async call. Anyway, I've reverted this change as wait_for_completion now defaults to false.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java]: The `cancel()` method is a bit trappy. I have removed the return value and made it idempotent in https://github.com/elastic/elasticsearch/pull/54312/commits/9a55fe41951f6bf0704ac5fe3690aa1206f832c5.

**Comment by dnhatn** [comment]
@ywelsch @javanna Thanks for reviewing. This is ready again.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/tasks/CancelTasksRequest.java
Comment (ywelsch) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/tasks/CancelTasksRequest.java:153]: Use this in `TasksRequestConvertersTests`?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/CancelTasksRequest.java:89]: call this `getWaitForCompletion`

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCancelTasksAction.java:70]: instead of introducing this constant, just use `cancelTasksRequest.getWaitForCompletion()` here

File: client/rest-high-level/src/main/java/org/elasticsearch/client/tasks/CancelTasksRequest.java
Comment (ywelsch) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/tasks/CancelTasksRequest.java:153]: As well as in `TasksClientDocumentationIT` to document it there as well?

**Review by ywelsch — APPROVED** [review]
I've left a few more nits, no need for another review, however. LGTM

**Comment by dnhatn** [comment]
Thanks Yannick.

**Pull request merged**
